histogram/src/lib.rs

```
@@ -1,206 +1,121 @@
-        histogram.increment(10000000, 1).unwrap();
-
-        let percentiles = histogram.percentiles(&[25.0, 75.0]).unwrap();
-
-        assert_eq!(histogram.percentile(25.0).map(|b| b.high()), Ok(1));
-        assert_eq!(histogram.percentile(75.0).map(|b| b.high()), Ok(10010623));
-
-        assert_eq!(percentiles.get(0).map(|b| b.bucket().high()), Some(1));
-        assert_eq!(
-            percentiles.get(1).map(|b| b.bucket().high()),
-            Some(10010623)
-        );
-    }
-
-    #[test]
-    fn increment_and_decrement() {
-        let histogram = Histogram::builder().build().unwrap();
-        assert_eq!(
-            histogram.percentile(0.0).map(|b| b.count()),
-            Err(Error::Empty)
-        );
-
-        histogram.increment(1, 1).unwrap();
-        assert_eq!(histogram.percentile(0.0).map(|b| b.count()), Ok(1));
-
-        histogram.decrement(1, 1).unwrap();
-        assert_eq!(
-            histogram.percentile(0.0).map(|b| b.count()),
-            Err(Error::Empty)
-        );
-    }
+        // a vec to hold the results
+        let mut result = Vec::with_capacity(percentiles.len());
+
+        let mut percentile_idx = 0;
+        let mut bucket_idx = 0;
+        let mut partial_sum = self.get_count(bucket_idx) as u128;
+
+        while percentile_idx < percentiles.len() {
+            let percentile = percentiles[percentile_idx];
+            let count = counts[percentile_idx];
+
+            // repeatedly push percentile-bucket pairs into the result while our
+            // current partial sum fulfills the count needed for each percentile
+            if partial_sum >= count {
+                result.push((percentile, self.get_bucket(bucket_idx).unwrap()));
+                percentile_idx += 1;
+                continue;
+            }
+
+            // increment the partial sum by the count of the next bucket
+            bucket_idx += 1;
+            if bucket_idx >= self.config().total_bins() {
+                break;
```

I think we only hit this case if a percentile > 100 is requested -- should we also filter those out while converting percentiles to counts?

---

We should return an error if percentile as f64 > 100.0 - in fact there is an error type defined for this, though I think the logic that enforces the constraints is missing...

---

Fixed.